sys/include/event.h

```diff
@@ -0,0 +1,218 @@
+ * into a queue.
+ *
+ * An event is a structure containing a pointer to an event handler. It can be
+ * extended to provide context or arguments to the handler.  It can also be
+ * embedded into existing structures (see examples).
+ *
+ * Compared to msg or mbox, this some fundamental differences:
+ *
+ * 1. events are "sender allocated". Unlike msg_send(), event_post() never
+ *    blocks or fails.
+ * 2. events contain everything necessary to handle them, thus a thread
+ *    processing the events of an event queue doesn't need to be changed in
+ *    order to support new event types.
+ * 3. events can be safely used (and actually perform best) when used within
+ *    one thread, e.g., in order to create a state-machine like process flow.
+ *    This is not (easily) possible using msg queues, as they might fill up.
+ * 4. an event can only be queued in one event queue at the same time.
+ *    Notifying many queues using only one event object is not possible with
+ *    this imlementation.
+ *
+ * At the core, event_wait() uses thread flags to implement waiting for events
+ * to be queued. Thus event queues can be used safely and efficiently in combination
+ * with thread flags and msg queues.
+ *
+ * Examples:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * // simple event handler
+ * static void handler(event_t *event)
+ * {
+ *    printf("triggered 0x%08x\n", (unsigned)event);
+ * }
+ *
+ * static event_t event = { .handler=handler };
+ * static event_queue_t queue;
+ *
+ * int main(void)
+ * {
+ *     event_queue_init(&queue);
+ *     event_loop(&queue);
+ * }
+ *
+ * [...] event_post(&queue, &event);
+ *
+ * // example for event extended event struct
+ * typedef struct {
+ *     event_t super;
+ *     const char *text;
+ * } custom_event_t;
+ *
+ * static void custom_handler(event_t *event)
+ * {
+ *     custom_event_t *custom_event = (custom_event_t *)event;
+ *     printf("triggered custom event with text: \"%s\"\n", custom_event->text);
+ * }
+ *
+ * static custom_event_t custom_event = { .super.callback=custom_handler, .text="CUSTOM EVENT" };
+ *
+ * [...] event_post(&queue, &custom_event)
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @{
+ *
+ * @file
+ * @brief       Event API
+ *
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ */
+
+#ifndef EVENT_H
+#define EVENT_H
+
+#include <stdint.h>
+
+#include "irq.h"
+#include "thread_flags.h"
+#include "clist.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef THREAD_FLAG_EVENT
+/**
+ * @brief   Thread flag use to notify available events in an event queue
+ */
+#define THREAD_FLAG_EVENT   (0x1)
+#endif
+
+/**
+ * @brief   event_queue_t static initializer
+ */
+#define EVENT_QUEUE_INIT    { .waiter=(thread_t *)sched_active_thread }
+
+/**
+ * @brief   event structure forward declaration
+ */
+typedef struct event event_t;
+
+/**
+ * @brief   event handler type definition
+ */
+typedef void(*event_handler_t)(event_t*);
```

Comment:
`s/void(/void (/` and `s/event_t\*/event_t \*/`

Reply:
yep